[HUDI-1072] Introduce REPLACE top level action

satishkotha opened the pull request:
## What is the purpose of the pull request
Add replace a top level action. Implement insert_overwrite operation on top of replace action

## Brief change log

1) All post commit actions work on top of HoodieCommitMetadata. Create HoodieReplaceMetadata to be subclass of HoodieCommitMetadata
2) Add insertOverwrite as new operation on HoodieWriteClient. insertOverwrite uses REPLACE action to mark all existing file groups as 'invalid'
3) Change archival to delete replaced files before archiving REPLACE action metadata

There are two other things that needs to be addressed:
1) Everywhere, we try to invoke 'getCommitsTimeline'/'filterCommits', we need to review and make sure caller can handle replace actions. OR create new methods and refactor all invocations of getCommitsTimeline call new methods.
2) FileSystemView#getAllFileGroups (and other methods in view) by default excludes file groups that have been replaced. We have to make sure callers can handle file groups that existed before, but have been replaced. (Example: say, there is a CompactionPlan that includes file id f1. f1 is replaced at a later instant. If compaction is run after replace, it wont be able to find corresponding file group. We need to make sure compaction can make progress instead of throwing errors here.) 

## Verify this pull request
This change added tests. Verified basic actions using quick start and docker setup. Added tests for insertOverwrite and FileSystemView changes. Need to add additional tests for backward compatibility. Also need to change integration tests to include insertOverwrite operations. Created followup tickets here https://issues.apache.org/jira/browse/HUDI-868

This is an example .hoodie folder from quick start setup:
-rw-r--r--  1 satishkotha  wheel  1933 Aug 27 16:39 20200827163904.commit
-rw-r--r--  1 satishkotha  wheel     0 Aug 27 16:39 20200827163904.commit.requested
-rw-r--r--  1 satishkotha  wheel  1015 Aug 27 16:39 20200827163904.inflight
-rw-r--r--  1 satishkotha  wheel  2610 Aug 27 16:39 20200827163927.replace
-rw-r--r--  1 satishkotha  wheel  1024 Aug 27 16:39 20200827163927.replace.inflight
-rw-r--r--  1 satishkotha  wheel     0 Aug 27 16:39 20200827163927.replace.requested
drwxr-xr-x  2 satishkotha  wheel    64 Aug 27 16:39 archived
-rw-r--r--  1 satishkotha  wheel   235 Aug 27 16:39 hoodie.properties



## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

satishkotha commented:
@vinothchandar @bvaradar FYI. There are few things that I'm not fully happy with. But would like to get initial feedback and get agreement on high level approach.

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieReplaceStat.java:
```diff
@@ -0,0 +1,71 @@
+public class HoodieReplaceStat extends HoodieWriteStat {
+
+  // records from the 'getFileId()' can be written to multiple new file groups. This list tracks all new fileIds
+  private List<String> newFileIds;
```
Why are we not tracking dropped fileIds ?

satishkotha replied:
its part of HoodieWriteStat#fileId

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/view/RemoteHoodieTableFileSystemView.java:
```diff
@@ -355,6 +357,18 @@ public RemoteHoodieTableFileSystemView(String server, int port, HoodieTableMetaC
   }
 
+  @Override
+  public Stream<String> getAllExcludeFileGroups(final String partitionPath) {
```
getAllExcludeFileGroups -> getReplacedFileGroups ?

satishkotha replied:
changed

bvaradar commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/FileSystemViewHandler.java:
```diff
@@ -284,6 +284,13 @@ private void registerFileSlicesAPI() {
       writeValueAsString(ctx, dtos);
     }, true));
 
+    app.get(RemoteHoodieTableFileSystemView.ALL_EXCLUDE_FILEGROUPS_FOR_PARTITION_URL, new ViewHandler(ctx -> {
```
ALL_EXCLUDE_FILEGROUPS_FOR_PARTITION_URL -> REPLACED_FILEGROUPS_FOR_PARTITION_URL

Lets use a single consistent name "replaced" instead of exclude everywhere.

satishkotha replied:
changed

bvaradar commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/HoodieCommonTestHarness.java:
```diff
@@ -104,4 +104,8 @@ protected SyncableFileSystemView getFileSystemViewWithUnCommittedSlices(HoodieTa
     return HoodieTableType.COPY_ON_WRITE;
   }
+
+  protected boolean areTimeTravelTestsEnabled() {
```
why is this needed ?

satishkotha replied:
RocksDB and Remote FileSystemViews are not implemented yet. So i temporarily disabled those tests. Moved this method to view tests instead of common

bvaradar commented on hudi-spark/src/main/java/org/apache/hudi/internal/HoodieDataSourceInternalWriter.java:
```diff
@@ -102,7 +102,7 @@ public void commit(WriterCommitMessage[] messages) {
 
     try {
-      writeClient.commitStats(instantTime, writeStatList, Option.empty());
+      writeClient.commitStats(instantTime, writeStatList, Option.empty(), HoodieTimeline.COMMIT_ACTION); //TODO get action type from HoodieWriterCommitMessage
```
Are you planning to address this TODO as part of this PR ?

satishkotha replied:
Fixed

bvaradar commented on hudi-timeline-service/src/main/java/org/apache/hudi/timeline/service/handlers/FileSliceHandler.java:
```diff
@@ -89,6 +89,10 @@ public FileSliceHandler(Configuration conf, FileSystemViewManager viewManager) t
         .collect(Collectors.toList());
   }
 
+  public List<String> getExcludeFileGroups(String basePath, String partitionPath) {
```
Rename to getReplacedFileGroups

satishkotha replied:
Done

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/view/SpillableMapBasedFileSystemView.java:
```diff
@@ -77,6 +79,13 @@ public SpillableMapBasedFileSystemView(HoodieTableMetaClient metaClient, HoodieT
+  protected Map<String, Set<String>> createPartitionToExcludeFileGroups() {
+    // TODO should we create another spillable directory under baseStoreDir?
+    // the exclude file group is expected to be small, so use parent class in-memory representation
+    return super.createPartitionToExcludeFileGroups();
```
Please follow the same structure like the one we did for compaction.

satishkotha replied:
Please take a look if i did it correctly. (To be honest, dont fully understand compaction implementation in great detail)

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/view/TableFileSystemView.java:
```diff
@@ -148,6 +148,9 @@
    */
   Stream<HoodieFileGroup> getAllFileGroups(String partitionPath);
 
+  Stream<String> getAllExcludeFileGroups(String partitionPath);
```
I think I have mentioned it somewhere else. Lets denote this feature with the consistent term "Replace" everywhere.

satishkotha replied:
Done

bvaradar commented:
@satishkotha : How  is incremental Reading handled.Are we going to support it ? If not, Are you going to throw exceptions ?  As we are cleaning up replaced files during archiving only but perform normal cleanup during cleaner operations, there will be cases when file versions of new files getting cleaned up before the old file versions getting removed.

bvaradar commented on hudi-client/src/main/java/org/apache/hudi/table/HoodieTimelineArchiveLog.java:
```diff
@@ -301,6 +304,44 @@ private void deleteAnyLeftOverMarkerFiles(JavaSparkContext jsc, HoodieInstant in
+
+  private void deleteFileGroups(String partitionPath, Set<String> fileIdsToDelete, HoodieInstant instant) {
+    try {
+      FileStatus[] statuses = metaClient.getFs().listStatus(FSUtils.getPartitionPath(metaClient.getBasePath(), partitionPath));
```
Instead of direct listStatus, can you use FileSystemViewAbstraction to get the file-group and then delete each files in it ? THis way, once consolidated metadata becomes available, you can take advantage of that. cc @prashantwason

satishkotha replied:
Modified to use FileSystemViews

satishkotha commented on hudi-common/src/main/java/org/apache/hudi/common/table/view/AbstractTableFileSystemView.java:
```diff
@@ -173,29 +180,59 @@ protected void refreshTimeline(HoodieTimeline visibleActiveTimeline) {
-        logFiles.get(pair).forEach(group::addLogFile);
-      }
+      String partitionPath = pair.getKey();
+      if (isExcludeFileGroup(partitionPath, fileId)) {
```
Modified. Please take a look. RocksDB needs to be implemented.

satishkotha commented:
@bvaradar I addressed your comments.I also created followup items here https://issues.apache.org/jira/browse/HUDI-868 as you suggested. 

 Please take another look when possible.

satishkotha commented:
@vinothchandar @bvaradar  Addressed most of your suggestions. Couple other followup items I need help from you on:

1) You suggested to remove HoodieReplaceStat.  I ran into minor implementation issue removing it. Basically, HoodieWriteClient operations return JavaRDD<WriteStatus>.  SparkSqlWriter uses these WriteStatus to create metadata (.commit/.replace etc).  Each WriteStatus comes with HoodieWriteStat (which is expected to be non-null in many places). This HoodieWriteStat is used for many post commit operations. So if we want to remove HoodieReplaceStat,  we can either
a) change signature of WriteClient operations to return a new structured object instead of just returning JavaRDD<WriteStatus>. This object would contain  JavaRDD<WriteStatus> for newly created files and List<HoodieFileGroupId> for tracking file groups replaced. We have to change post commit operations to look at this new object instead of WriteStatus.
OR
b) Return a WriteStatus for replaced file groups too. WriteClient operations can continue to return JavaRDD<WriteStatus>. Each WriteStatus has HoodieWriteStat which can be a token value (null?) for replaced file groups. 

Either way, this is somewhat involved change, so would like to get your feedback before starting implementation. What do you guys think?

2) Deleting replaced file groups during archival vs clean. I've this deletion logic implemented in archival per our earlier conversation. But, as I mentioned, this may lead to storage inefficiency. For example, a) clean retain is set to 1 commit.  b) archival is done after 24 commits. We keep all the data for replaced files until archival happens. 

Let me know if you guys have any other comments.

vinothchandar commented:
>You suggested to remove HoodieReplaceStat

I think the suggestion was to simplify HoodieReplaceMetadata such that it only contains the extra information about replaced file groups. and use the HoodieCommitMetadata and its HoodieWriteStat for tracking the new file groups written.
We could have HoodieReplaceStat to be part of the WriteStatus itself for tracking the additional information about replaced file groups? 

On cleaning vs archival, it would be good if we can implement this in cleaning. But can that be a follow-on item? Practically speaking, typical deployments don't configure cleaning that low.

satishkotha commented:
@vinothchandar As discussed, i added boolean in WriteStatus and removed HoodieReplaceStat. See this [diff](https://github.com/apache/hudi/pull/2048/commits/94b275dbd20ec82ebe568b47bb28447d92ab996f). I committed it as a separate git-sha because this still looks  somewhat awkward IMO. Please take a look and I can revert or reimplement in a different way

Also, created https://issues.apache.org/jira/browse/HUDI-1276 for cleaning replaced file during clean.

I also renamed 'replace' to 'replacecommit' everywhere as you suggested. 

Please let me know if you have additional comments/suggestions

vinothchandar reviewed (commented):
Just getting started

vinothchandar commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/HoodieTestUtils.java:
```diff
@@ -200,6 +199,14 @@ public static void createInflightCommitFiles(String basePath, String... instantT
     }
   }
 
+  public static HoodieWriteStat createReplaceStat(final String partitionPath, final String fileId1) {
```
there is nothing specific about replace in this method? should we move this to the test class itself. inline?

satishkotha replied:
Done.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieTimeline.java:
```diff
@@ -133,6 +137,21 @@
+   *
+   * @return
+   */
+  HoodieDefaultTimeline getWriteActionTimeline();
```
see earlier comment on whether we need this method.

vinothchandar replied:
To expand, I am wondering if we should just include replacecommit within `getCommitsAndCompactionTimeline()`. Most of its callers are around compaction/savepoint/restore etc. So we may not be seeing some cases here. 

Things work for now, since filterCompletedInstants() etc are including replace commit in the timeline when filtering for queries. Semantically, if replace is a commit level action that can add new data to the timeline, then we should just treat it like delta commit IMO. 

Would anything break if we did do that?

satishkotha replied:
I removed and included getCommitsAndCompactionTimeline. I think we will run into some edge cases for MOR tables where something would break. But don't have concrete examples. We can run it for a while on MOR table and see if works.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/view/RocksDbBasedFileSystemView.java:
```diff
@@ -371,6 +371,47 @@ void removeBootstrapBaseFileMapping(Stream<BootstrapBaseFileMapping> bootstrapBa
   }
 
+  @Override
+  protected void resetReplacedFileGroups(final Map<HoodieFileGroupId, HoodieInstant> replacedFileGroups) {
```
@bvaradar if you can take a pass at these, that would be great

vinothchandar replied:
General question I have around incremental file system view and rocksDB like persistent file system view storage is whether we will keep this list updated. i.e when the archival/cleaning runs, how do we ensure the deleted replaced file groups are no longer tracked inside rocksdb. 

I guess the lines below, are doing a bulk delete and insert to achieve the same?

satishkotha replied:
Yes,  we do delete and insert. Definitely, would be helpful to have someone with more experience review this.

bvaradar replied:
This part looks good. Going through rest of the code

vinothchandar commented on hudi-spark/src/main/java/org/apache/hudi/internal/HoodieDataSourceInternalWriter.java:
```diff
@@ -102,7 +104,9 @@ public void commit(WriterCommitMessage[] messages) {
 
     try {
-      writeClient.commitStats(instantTime, writeStatList, Option.empty());
+      writeClient.commitStats(instantTime, writeStatList, Option.empty(),
```
better approach for these situations generally is to introduce a `commitStats(.)` that does not take the last argument and deal with it internally inside WriteClient. 

This way the code will remain more readable, without having to reference replace stats in bulk_insert, which have nothing to do with each other. hope that makes sense

satishkotha replied:
Sure. Added new method

vinothchandar commented on hudi-spark/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala:
```diff
@@ -111,6 +111,9 @@ private[hudi] object HoodieSparkSqlWriter {
       }
 
+      val metaClient = new HoodieTableMetaClient(sparkContext.hadoopConfiguration, path.get)
+      val commitActionType = DataSourceUtils.getCommitActionType(operation, metaClient)
```
should we have a better way of getting the commit action type? I am bit concerned about creating new metaClient just for this.

satishkotha replied:
We just need tableType to create commit/deltacommit in the default case. I moved this to a static method in util class and removed dependency on MetaClient.

vinothchandar commented on hudi-spark/src/main/scala/org/apache/hudi/HoodieSparkSqlWriter.scala:
```diff
@@ -376,31 +379,35 @@ private[hudi] object HoodieSparkSqlWriter {
   }
 
+  /**
+   * Scala says method cannot have more than 7 arguments. So group all table/action specific information into a case class.
```
please remove the scala specific comment. its good to encapsulate like this anyway

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/table/HoodieTable.java:
```diff
@@ -213,6 +213,12 @@ public abstract HoodieWriteMetadata insertPrepped(JavaSparkContext jsc, String i
+  /**
+   * Logically delete all existing records and Insert a batch of new records into Hoodie table at the supplied instantTime.
+   */
+  public abstract HoodieWriteMetadata insertOverwrite(JavaSparkContext jsc, String instantTime,
```
On second thoughts, I am okay leaving this as-is for now as well. and reeval when acutally implementing clustering

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/table/HoodieTimelineArchiveLog.java:
```diff
@@ -301,6 +304,61 @@ private void deleteAnyLeftOverMarkerFiles(JavaSparkContext jsc, HoodieInstant in
+    TableFileSystemView fileSystemView = this.table.getFileSystemView();
+    ensureReplacedPartitionsLoadedCorrectly(instant, fileSystemView);
+
+    Stream<HoodieFileGroup> fileGroupsToDelete = fileSystemView
```
One more consideration, as I went through the remainder of the PR. if there was an pending compaction for the replaced file group, then the file group metadata we encode may miss new base files produced as a result of the compaction. This scenario needs to be thought thru.

satishkotha replied:
Yes, compaction is the primary reason I only recorded fileId in the replace metadata. When deleting, we can get all file paths (through view or by listing using consolidated metadata) that have same fileId and delete these files.  

There can be race conditions that compaction might create a new file with replaced fileId after we queried for existing files though. But because FileSystemView#get methods do not include replaced file groups, I think this is unlikely to happen. I'm not sure if there are edge cases with long running compactions.

Please suggest any other improvements.

satishkotha replied:
So, this is tricky to explain. In FileSystemView, only metadata seems to be eagerly loaded. file groups are not eagerly loaded. i.e., fetchAllStoredFileGroups() returns empty.  For replace instants, we need to get List<FileSlice> for all fileId. Because fetchAllStoredFileGroups() is empty, its also returning empty list of FileSlices. So we dont delete replaced files.

I think instead of creating new HoodieTable in constructor. passing that from callers would help workaround this problem. But that is somewhat involved change because of test dependencies. Also, it might be better to refresh partition content in case new files are created by compaction or other process and somehow that is not reflected in table views. This might be safer option.

Let me know if you want me to work on passing in HoodieTable to HoodieTimelineArchiveLog constructor.

bvaradar replied:
@satishkotha Here is the plan as we discussed.
1. Change the signature of fileSystemView.getReplacedFileGroupsBeforeOrOn to also take in partitionId
2.  In HoodieTimelineArchiveLog.deleteReplacedFileGroups, read the replace metadata (which we are already doing) and for each partition, call fileSystemView.getReplacedFileGroupsBeforeOrOn().
3. (2) must be done in such a way that we are calling the fileSystemView.getReplacedFileGroupsBeforeOrOn in parallel.

This should allow for lazy loading semantics to be retained at file-system view.

satishkotha replied:
Done. PTAL.

vinothchandar reviewed (commented):
Did a detailed review. There are some involved comments, but we are nearing the finish line. 

Thoughts on whether the callers of timeline need to see replacecommi or not. 
I feel we should treat it just like delta commit. The piece of code that needs to see the replaced file groups is the archival process. Otherwise, everything else can just work with the replaced file groups. It could be that compaction is scheduled for a file group that was replaced, and we should just ensure it can complete successfully with this change (my understanding is that it should, given we materialize the compaction plan upfront before replace is triggered). This is a scenario we cannot avoid in real world, so we can just embrace it and build towards that.  It will make clustering support so much more simpler to do as well. 

Finally, would be good to understand how this will interplay with  consolidated metadata. I reckon we need to issue deletes for the replaced file groups from timeline archive log (or) preferrably the cleaning code path

vinothchandar commented:
@satishkotha can you please resolve all the comments, that you have addressed already. That way we can track whats pending

satishkotha commented on hudi-client/src/main/java/org/apache/hudi/table/HoodieTimelineArchiveLog.java:
```diff
@@ -301,6 +304,61 @@ private void deleteAnyLeftOverMarkerFiles(JavaSparkContext jsc, HoodieInstant in
+  private boolean deletePath(Path path, HoodieInstant instant) {
+    try {
+      LOG.info("Deleting " + path + " before archiving " + instant);
+      metaClient.getFs().delete(path);
```
metaclient.getFs() seems to be following singleton pattern, so it doesnt seem expensive to get this. am i reading it incorrectly?

I can work to setup parallel deletes

satishkotha replied:
for parallel deletes, JavaSparkContext is not exposed to Archive process. Since we anyway want to move this to be part of clean, is it ok if  I defer this to https://issues.apache.org/jira/browse/HUDI-1276?

satishkotha replied:
i missed that we already have JavaSparkContext. Implemented parallel clean up. PTAL

satishkotha commented on hudi-client/src/main/java/org/apache/hudi/table/HoodieTimelineArchiveLog.java:
```diff
@@ -301,6 +304,61 @@ private void deleteAnyLeftOverMarkerFiles(JavaSparkContext jsc, HoodieInstant in
+    Option<HoodieInstant> replaceInstantOption = metaClient.getActiveTimeline().getCompletedAndReplaceTimeline()
+        .filter(replaceInstant -> replaceInstant.getTimestamp().equals(instant.getTimestamp())).firstInstant();
+
+    replaceInstantOption.ifPresent(replaceInstant -> {
```
correct, i'll restructure this code

satishkotha commented on hudi-client/src/main/java/org/apache/hudi/table/HoodieTimelineArchiveLog.java:
```diff
@@ -301,6 +304,61 @@ private void deleteAnyLeftOverMarkerFiles(JavaSparkContext jsc, HoodieInstant in
   }
 
+  private void deleteReplacedFiles(HoodieInstant instant) {
+    if (!instant.isCompleted()) {
```
correct. will add it.

satishkotha replied:
Implemented parallel deletion. PTAL.

satishkotha commented on hudi-client/src/main/java/org/apache/hudi/client/HoodieWriteClient.java:
```diff
@@ -576,7 +592,8 @@ public String startCommit() {
     }
     String instantTime = HoodieActiveTimeline.createNewInstantTime();
-    startCommit(instantTime);
+    HoodieTableMetaClient metaClient = createMetaClient(true);
```
sure

satishkotha commented on hudi-client/src/main/java/org/apache/hudi/client/HoodieWriteClient.java:
```diff
@@ -586,15 +603,23 @@ public String startCommit() {
-  private void startCommit(String instantTime) {
+  private void startCommit(String instantTime, String actionType) {
     LOG.info("Generate a new instant time " + instantTime);
     HoodieTableMetaClient metaClient = createMetaClient(true);
```
Good point. Will do.

satishkotha commented on hudi-client/src/main/java/org/apache/hudi/table/action/HoodieWriteMetadata.java:
```diff
@@ -94,6 +94,14 @@ public void setWriteStats(List<HoodieWriteStat> writeStats) {
     this.writeStats = Option.of(writeStats);
   }
 
+  public Option<List<HoodieWriteStat>> getReplacetats() {
```
This is not needed anymore given we are tracking replaced files as boolean in WriteStatus. I removed this.

satishkotha commented on hudi-client/src/main/java/org/apache/hudi/table/action/commit/InsertOverwriteCommitActionExecutor.java:
```diff
@@ -0,0 +1,107 @@
+
+  private Stream<String> getAllExistingFileIds(String partitionPath) {
+    // because new commit is not complete. it is safe to mark all base files as old files
+    return table.getBaseFileOnlyView().getAllBaseFiles(partitionPath).map(baseFile -> baseFile.getFileId());
```
My bad, missed that insert into log files case. I fixed it now. thanks for finding this bug.

satishkotha commented on hudi-common/src/main/java/org/apache/hudi/common/model/ActionType.java:
```diff
@@ -22,5 +22,5 @@
  */
 public enum ActionType {
-  commit, savepoint, compaction, clean, rollback
+  commit, savepoint, compaction, clean, rollback, replacecommit
```
Filed HUDI-1281

satishkotha commented on hudi-common/src/test/java/org/apache/hudi/common/testutils/HoodieTestDataGenerator.java:
```diff
@@ -378,9 +398,9 @@ public static void createCompactionAuxiliaryMetadata(String basePath, HoodieInst
     FileSystem fs = FSUtils.getFs(basePath, configuration);
     try (FSDataOutputStream os = fs.create(commitFile, true)) {
-      HoodieCompactionPlan workload = new HoodieCompactionPlan();
+      HoodieCompactionPlan workload = HoodieCompactionPlan.newBuilder().setVersion(1).build();
```
Version is not being set by default, so reading test plan generated here is failing. So I explicitly set version.

satishkotha commented on hudi-common/src/main/java/org/apache/hudi/common/util/CommitUtils.java:
```diff
@@ -0,0 +1,82 @@
+
+  private static final Logger LOG = LogManager.getLogger(CommitUtils.class);
+
+  public static HoodieCommitMetadata buildWriteActionMetadata(List<HoodieWriteStat> writeStats,
```
Added

satishkotha commented on hudi-common/src/main/avro/HoodieArchivedMetaEntry.avsc:
```diff
@@ -36,6 +36,14 @@
          ],
          "default": null
       },
+      {
```
Moved it. will check if its possible to add unit test for backward compatibility.

satishkotha commented on hudi-common/src/main/avro/HoodieReplaceCommitMetadata.avsc:
```diff
@@ -0,0 +1,64 @@
+   "name":"HoodieReplaceCommitMetadata",
+   "fields":[
+      {
+         "name":"partitionToWriteStats",
```
Moved partitionToReplaceStats at the end. Its also possible to make this reference HoodieCommitMetadata directly i.e.,

HoodieReplaceCommitMetadata contains HoodieCommitMetadata and replaceFileIds. But converting between json replaceCommit and avro version will require another layer of transform code. So to keep it simple, I copied all the fields from json structure.

satishkotha commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieReplaceCommitMetadata.java:
```diff
@@ -0,0 +1,137 @@
+@JsonIgnoreProperties(ignoreUnknown = true)
+public class HoodieReplaceCommitMetadata extends HoodieCommitMetadata {
+  private static final Logger LOG = LogManager.getLogger(HoodieReplaceCommitMetadata.class);
+  protected Map<String, List<HoodieWriteStat>> partitionToReplaceStats;
```
I changed it to List<String> to include only fileIds. I'm inclined against storing all file slices because they can evolve between metadata creation and archival/clean. Let me know if this understanding is incorrect

satishkotha commented on hudi-client/src/test/java/org/apache/hudi/client/TestHoodieClientOnCopyOnWriteStorage.java:
```diff
@@ -880,6 +880,89 @@ public void testDeletesWithDeleteApi() throws Exception {
+  /**
+   * Verify data in parquet files matches expected records and commit time.
+   */
+  private void verifyParquetFileData(String commitTime, List<HoodieRecord> expectedRecords, List<WriteStatus> allStatus) {
```
renamed it for now. I'll look into if there are any other helpers for doing this.

satishkotha commented on hudi-client/src/main/java/org/apache/hudi/table/action/commit/InsertOverwriteCommitActionExecutor.java:
```diff
@@ -0,0 +1,107 @@
+  @Override
+  protected JavaRDD<WriteStatus> processInputRecords(JavaRDD<HoodieRecord<T>> inputRecordsRDD, WorkloadProfile profile) {
+    // get all existing fileIds to mark them as replaced
+    JavaRDD<WriteStatus> replaceStatuses = getAllReplaceWriteStatus(profile);
```
I refactored it and removed boolean from WriteStatus. PTAL

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/view/IncrementalTimelineSyncFileSystemView.java:
```diff
@@ -251,6 +262,28 @@ private void addRollbackInstant(HoodieTimeline timeline, HoodieInstant instant)
+   * @param timeline Hoodie Timeline
+   * @param instant REPLACE Instant
+   */
+  private void addReplaceInstant(HoodieTimeline timeline, HoodieInstant instant) throws IOException {
```
Also, can you also add test cases for incremental file-system view for both addReplaceInstant and removeReplaceInstant in TestIncrementalFSViewSync ?

satishkotha replied:
I need to understand this flow a bit more. But, have a question on why we need to track commit-action-type and timestamp. Today, HoodieRollbackMetadata tracks successFiles, deletedFiles etc.  Do you think we can add replacedFileIds also there? This will be empty for regular commits. But for replace commits, it will have some content.  If this value is present, we can remove corresponding fileIds from View#replacedFileGroups. Let me know if i'm missing anything with this approach.

bvaradar replied:
Discussed with @satishkotha. We will track commit action type along with instant since we have introduced one another type of "commit" -> replace. Restore can perform custom handling only for this action type. It is safer to let Incremental file-system view to revert replace mappings (in memory or rocksdb) by providing the replace instant time as it  isrobust in case of partial restore failures.

satishkotha replied:
@bvaradar Made the change and added basic test. Please take a look. If the general approach looks good. I'll add more complex tests.

satishkotha commented on hudi-client/src/main/java/org/apache/hudi/table/HoodieTimelineArchiveLog.java:
```diff
@@ -301,6 +304,61 @@ private void deleteAnyLeftOverMarkerFiles(JavaSparkContext jsc, HoodieInstant in
+    Stream<HoodieFileGroup> fileGroupsToDelete = fileSystemView
+        .getReplacedFileGroupsBeforeOrOn(instant.getTimestamp());
+
+    fileGroupsToDelete.forEach(fg -> {
```
Implemented parallel execution. PTAL

bvaradar commented:
@satishkotha : Please ping me in the PR  when you have updates and I can give incremental comments if needed.

satishkotha commented:
> @satishkotha : Please ping me in the PR when you have updates and I can give incremental comments if needed.

@bvaradar IncrementalTimeline resotre is the only big pending item. I'll get to it in later part of this week.

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/view/AbstractTableFileSystemView.java:
```diff
@@ -738,7 +799,9 @@ private String formatPartitionKey(String partitionStr) {
   Stream<FileSlice> fetchLatestFileSliceInRange(List<String> commitsToReturn) {
-    return fetchAllStoredFileGroups().map(fileGroup -> fileGroup.getLatestFileSliceInRange(commitsToReturn))
+    return fetchAllStoredFileGroups()
+        .filter(fileGroup -> !isFileGroupReplacedBeforeAny(fileGroup, commitsToReturn))
```
@satishkotha  : As discussed, All the replace filtering needs to move to getXXX() apis as the fetch APIs are only responsible for fetching file slices/base-files from different types of storage.

satishkotha replied:
Changed. Please take a look.

bvaradar commented on hudi-common/src/main/avro/HoodieRestoreMetadata.avsc:
```diff
@@ -34,6 +34,8 @@
-     }
+     },
+     /* overlaps with 'instantsToRollback' field. Adding this to track action type for all the instants being rolled back. */
+     {"name": "restoreInstantInfo", "default": null, "type": {"type": "array", "default": null, "items": ["null", "HoodieInstantInfo"]}}
```
Can you use pretty print mode (multi-lines) in the same way other section is presented.

bvaradar reviewed (changes requested):
@satishkotha : Please go ahead and add unit-tests.

satishkotha commented on hudi-client/src/main/java/org/apache/hudi/client/HoodieWriteClient.java:
```diff
@@ -586,24 +602,39 @@ public String startCommit() {
    * @param instantTime Instant time to be generated
    */
   public void startCommitWithTime(String instantTime) {
+    HoodieTableMetaClient metaClient = createMetaClient(true);
```
This is not calling function in the next line (calling method after that). So we only create meta client once. Please double check and let me know if i'm misinterpreting your suggestion.

bvaradar replied:
My bad. Got confused with the method naming :)

bvaradar commented on hudi-client/src/main/java/org/apache/hudi/client/HoodieWriteClient.java:
```diff
@@ -586,24 +602,39 @@ public String startCommit() {
-  private void startCommit(String instantTime) {
-    LOG.info("Generate a new instant time " + instantTime);
-    HoodieTableMetaClient metaClient = createMetaClient(true);
+  private void startCommit(String instantTime, String actionType, HoodieTableMetaClient metaClient) {
```
Does this have to be non-static ? Can it be moved to CommitUtils ?

satishkotha replied:
This is a private method. Do you want to make this public static?  Personally, I think having all startCommit methods in HoodieWriteClient makes more sense because user workflow is

1) writeClient#startCommit
2) writeClient#upsert
3) writeClient#commit

But if you have a strong preference to make this part of CommitUtils, I can move it. let me know.

bvaradar replied:
Lets leave it for now If needed, we can refactor later.

bvaradar commented:
@satishkotha : when the conflicts are resolved and we can land this.

bvaradar reviewed (approved):
Thanks @satishkotha  Looks good assuming the follow-up tasks : https://issues.apache.org/jira/browse/HUDI-1042
